src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,469 @@
+    # each subclass must provide.
+    # If you need additional information, this class follows the "Builder" pattern. Use the
+    # "builder" methods to attach additional information.
+    # NB: We define `__new__` rather than `__init__` because otherwise subclasses would have to as
+    # well
+    def __new__(
+        cls,
+        *flag_names: str,
+        default: PropType,
+        option_type: Type | None = None,
+        help: str,
+    ):
+        self = super().__new__(cls)
+        self.flag_names = flag_names
+        if option_type is None:
+            option_type = cls.option_type  # type: ignore[attr-defined]
+
+        self.kwargs = dict(
+            type=option_type,
+            default=default,
+            help=help,
+        )
+        return self
+
+    @overload
+    def __get__(self, obj: None, objtype: Any) -> Option:
+        ...
+
+    @overload
+    def __get__(self, obj: object, objtype: Any) -> PropType:
+        ...
+
+    def __get__(self, obj, objtype):
+        if obj is None:
+            return self
+        long_name = self.flag_names[-1]
+        option_value = getattr(obj.options, long_name[2:].replace("-", "_"))
+        return self._convert_(option_value)
+
+    # Subclasses can override if necessary
+    def _convert_(self, val: Any) -> PropType:
+        if val is None:
+            # NB: This will only ever be the case if PropType is an Optional.
+            return None  # type: ignore[return-value]
+        return cast("PropType", self.kwargs["type"](val))
+
+    # ===== "Builder" Methods =====
+    def advanced(self: OptionT) -> OptionT:
+        self.kwargs["advanced"] = True
+        return self
+
+    def metavar(self: OptionT, metavar: str, /) -> OptionT:
+        self.kwargs["metavar"] = metavar
+        return self
+
+    def mutually_exclusive_group(self: OptionT, mutually_exclusive_group: str, /) -> OptionT:
+        self.kwargs["mutually_exclusive_group"] = mutually_exclusive_group
+        return self
+
+    def default_help_repr(self: OptionT, default_help_repr: str, /) -> OptionT:
+        self.kwargs["default_help_repr"] = default_help_repr
+        return self
+
+    def will_be_removed(self: OptionT, *, version: str, hint: str) -> OptionT:
+        self.kwargs["removal_version"] = version
+        self.kwargs["removal_hint"] = hint
+        return self
+
+    # @TODO: These only seem relevant on global options?
+    def daemoned(self: OptionT) -> OptionT:
+        self.kwargs["daemon"] = True
+        return self
+
+    def non_fingerprinted(self: OptionT) -> OptionT:
+        self.kwargs["fingerprint"] = False
+        return self
+
+
+class StrOption(Option[StrT]):
+    option_type: Type[str] = str
+
+    @overload
+    def __new__(cls, *flag_names: str, default: str, help: str) -> StrOption[str]:
+        ...
+
+    @overload
+    def __new__(cls, *flag_names: str, help: str) -> StrOption[str | None]:
+        ...
+
+    def __new__(cls, *flag_names, default=None, help):
```

Comment:
This is the correctly-typed way to teach type-checkers what the generic type of yourself is based on provided argument types.

NOTE: Some places may suggest using `__init__` and annotating `self`, however that is not standardized by any PEP and support varies by static type checker.

Reply:
I think the final `__new__` should have type hints, right? `default: str | None = None`. Ditto every concrete class.

Reply:
The final one is ignored completely by type-checkers as per PEP484

> The `@overload`-decorated definitions are for the benefit of the type checker only, since they will be overwritten by the non-`@overload`-decorated definition, while the latter is used at runtime but should be ignored by a type checker.